public/app/plugins/datasource/prometheus/result_transformer.ts

```diff
@@ -37,6 +41,94 @@ interface TimeAndValue {
   [TIME_SERIES_VALUE_FIELD_NAME]: number;
 }
 
+// V2 result trasnformer used to transform query results from queries that were run trough prometheus backend
+export function transformV2(response: DataQueryResponse, options: DataQueryRequest<PromQuery>) {
+  // Get refIds that have table format as we need to process those to table reuslts
+  const tableRefIds = options.targets.filter((target) => target.format === 'table').map((target) => target.refId);
+  const [tableResults, otherResults] = partition(response.data, (dataFrame) =>
+    dataFrame.refId ? tableRefIds.includes(dataFrame.refId) : false
+  );
+
+  // For table results, we need to transform data frames to table data frames
+  const tableFrames = tableResults.map((dataFrame) => {
+    const df = transformDFoTable(dataFrame, options.targets.length);
+    return df;
+  });
+
+  // Everything else is processed as time_series result and graph preferredVisualisationType
+  const otherFrames = otherResults.map((dataFrame) => {
+    const df = {
+      ...dataFrame,
+      meta: {
+        ...dataFrame.meta,
+        preferredVisualisationType: 'graph',
+      },
+    };
+    return df;
+  });
+
+  // Order by refIds
+  const orderedFrames = [...otherFrames, ...tableFrames].sort((df1, df2) => {
+    if (df1.refId && df2.refId) {
+      if (df1.refId < df2.refId) {
+        return -1;
+      }
+    }
+    return 1;
+  });
+
+  return { ...response, data: orderedFrames };
+}
+
+export function transformDFoTable(df: DataFrame, responseLength: number): DataFrame {
+  if (!df || df.length === 0) {
+    return df;
+  }
```

Comment:
is `!df` needed here given it can't be falsy?

i see that where it is used it's of type any, I guess this check is here because of that? in that case, wouldn't It make sense to move the check there?

Reply:
Yes, updated: https://github.com/grafana/grafana/pull/39133/commits/8b89a46ffc5bd2d6a283ecaaa2dfa15a21265877